create_run_args copy-by-ref fix

ankona opened the pull request:
Add a deep copy to the batch & run settings `__init__` methods. This ensures that changes to collections after pass-by-ref are not unexpectedly modified after creating run/batch settings instances.

MattToast commented on smartsim/settings/base.py:
```diff
@@ -78,19 +77,39 @@ def __init__(
-        self.run_args = init_default({}, run_args, dict)
-        self.env_vars = init_default({}, env_vars, dict)
+        self.exe = [exe] if container else [expand_exe_path(exe)]
+        self.exe_args = exe_args
```
Similar "expected `T` but got `T | None`"  for `run_args` and `env_vars`

ankona replied:
Fixed

MattToast commented on smartsim/settings/base.py:
```diff
@@ -78,19 +77,39 @@ def __init__(
+        return self._run_args
+    
+    @run_args.setter
+    def run_args(self, value: t.List[str]) -> None:
```
same guess for the `env_vars` setter

ankona replied:
copypasta monster fail. tyvm. fixed.

MattToast reviewed (approved):
LGTM!! A couple of small typing, but other wise looks great!

Approving in advance!!

ankona commented on smartsim/settings/base.py:
```diff
@@ -477,11 +496,14 @@ def set(self, arg: str, value: t.Optional[str] = None, condition: bool = True) -
-    def _set_exe_args(self, exe_args: t.Optional[t.Union[str, t.List[str]]]) -> t.List[str]:
+    @staticmethod
+    def _build_exe_args(exe_args: t.Optional[t.Union[str, t.List[str]]]) -> t.List[str]:
+        """ Convert exe_args input to a desired collection format"""
```
removed extraneous spacing

mellis13 reviewed (approved):
One small question/comment -- approving in advance.  Thanks!

ankona commented on smartsim/settings/base.py:
```diff
@@ -36,12 +37,10 @@ class RunSettings:
         self,
         exe: str,
-        exe_args: t.Optional[t.Union[str, t.List[str]]] = None,
+        exe_args: t.Optional[t.List[str]] = None,
```
I've undone the removal of the str as an option for setting exe_args.

ankona commented:
@mellis13 

I've rebased off the latest develop branch (internal typehints were added) and did some tweaks to come inline with expectations. Mainly, the dictionaries for settings are changed to `t.Dict[str, *t.Optional*[str]]` (optional values).

I also updated some tests and functions to handle the automatic stringification of configuration values when stored (e.g. internally, all settings are t.Dict[str, str] even if you want something to be an integer, like cpus_per_node). Many tests existed that pass a string in and expected a ValueError to be thrown. You'll see some unsightly `str(int(value))` to ensure those same errors occurred.

MattToast commented on smartsim/settings/slurmSettings.py:
```diff
@@ -201,7 +201,7 @@ def set_memory_per_node(self, memory_per_node: int) -> None:
         :type memory_per_node: int
         """
-        self.run_args["mem"] = f"{int(memory_per_node)}M"
+        self.run_args["mem"] = f"{str(int(memory_per_node))}M"
```
Is the str cast needed here to satisfy the typechecker? Should be redundant as it is cast implicitly by the f-str

ankona replied:
doh!

ankona replied:
reverted these changes

MattToast reviewed (approved):
LGTM! One last question, but nothing worth holding approval over
